# ИДЗ  №2

Сформулируйте реальную задачу, решение которой сводится к проверке статистической гипотезы о независимости двух случайных величин, измеренных в номинальной шкале.
- Опишите данные (дайте ссылку на источник данных).
- Сформулируйте статистическую задачу. Изложите критерий хи-квадрат и известные вам меры связи номинальных показателей.
- Примените к  данным критерий хи-квадрат. Вычислите указанные в пункте  Б) меры.
- Сделайте выводы.

In [37]:
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt 
import numpy as np

import kagglehub
from kagglehub import KaggleDatasetAdapter
from scipy.stats import chi2

import warnings
warnings.filterwarnings('ignore')

# Загрузка исходных данных с Kaggle

In [2]:
df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "sooyoungher/smoking-drinking-dataset",
  "smoking_driking_dataset_Ver01.csv"
)

df.head()

,sex,age,height,weight,waistline,sight_left,sight_right,hear_left,hear_right,SBP,...,LDL_chole,triglyceride,hemoglobin,urine_protein,serum_creatinine,SGOT_AST,SGOT_ALT,gamma_GTP,SMK_stat_type_cd,DRK_YN
0,Male,35,170,75,90.0,1.0,1.0,1.0,1.0,120.0,...,126.0,92.0,17.1,1.0,1.0,21.0,35.0,40.0,1.0,Y
1,Male,30,180,80,89.0,0.9,1.2,1.0,1.0,130.0,...,148.0,121.0,15.8,1.0,0.9,20.0,36.0,27.0,3.0,N
2,Male,40,165,75,91.0,1.2,1.5,1.0,1.0,120.0,...,74.0,104.0,15.8,1.0,0.9,47.0,32.0,68.0,1.0,N
3,Male,50,175,80,91.0,1.5,1.2,1.0,1.0,145.0,...,104.0,106.0,17.6,1.0,1.1,29.0,34.0,18.0,1.0,N
4,Male,50,165,60,80.0,1.0,1.2,1.0,1.0,138.0,...,117.0,104.0,13.8,1.0,0.8,19.0,12.0,25.0,1.0,N


In [24]:
print('Описание численных типов данных')
display(df.describe())
print('-'*10, '\n')
print('Описание категориальных типов данных')
display(df.describe(include='object'))

Описание численных типов данных


,age,height,weight,waistline,sight_left,sight_right,hear_left,hear_right,SBP,DBP,...,HDL_chole,LDL_chole,triglyceride,hemoglobin,urine_protein,serum_creatinine,SGOT_AST,SGOT_ALT,gamma_GTP,SMK_stat_type_cd
count,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,...,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000,991346.000000
mean,47.614491,162.240625,63.284050,81.233358,0.980834,0.978429,1.031495,1.030476,122.432498,76.052627,...,56.936800,113.037692,132.141751,14.229824,1.094224,0.860467,25.989308,25.755051,37.136347,1.608122
std,14.181339,9.282957,12.514241,11.850323,0.605949,0.604774,0.174650,0.171892,14.543148,9.889365,...,17.238479,35.842812,102.196985,1.584929,0.437724,0.480530,23.493386,26.308599,50.424153,0.818507
min,20.000000,130.000000,25.000000,8.000000,0.100000,0.100000,1.000000,1.000000,67.000000,32.000000,...,1.000000,1.000000,1.000000,1.000000,1.000000,0.100000,1.000000,1.000000,1.000000,1.000000
25%,35.000000,155.000000,55.000000,74.100000,0.700000,0.700000,1.000000,1.000000,112.000000,70.000000,...,46.000000,89.000000,73.000000,13.200000,1.000000,0.700000,19.000000,15.000000,16.000000,1.000000
50%,45.000000,160.000000,60.000000,81.000000,1.000000,1.000000,1.000000,1.000000,120.000000,76.000000,...,55.000000,111.000000,106.000000,14.300000,1.000000,0.800000,23.000000,20.000000,23.000000,1.000000
75%,60.000000,170.000000,70.000000,87.800000,1.200000,1.200000,1.000000,1.000000,131.000000,82.000000,...,66.000000,135.000000,159.000000,15.400000,1.000000,1.000000,28.000000,29.000000,39.000000,2.000000
max,85.000000,190.000000,140.000000,999.000000,9.900000,9.900000,2.000000,2.000000,273.000000,185.000000,...,8110.000000,5119.000000,9490.000000,25.000000,6.000000,98.000000,9999.000000,7210.000000,999.000000,3.000000


---------- 

Описание категориальных типов данных


,sex,DRK_YN
count,991346,991346
unique,2,2
top,Male,N
freq,526415,495858


## Типы данных в датасете и проврка на пустые значения 

In [33]:
print('Типы данных в датасете \n')
display(df[['sex', 'SMK_stat_type_cd']].info())
print('-'*10, '\n')
print('Проверка на пустые значения \n')
display(df[['sex', 'SMK_stat_type_cd']].isna().sum())

Типы данных в датасете 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 991346 entries, 0 to 991345
Data columns (total 2 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   sex               991346 non-null  object 
 1   SMK_stat_type_cd  991346 non-null  float64
dtypes: float64(1), object(1)
memory usage: 15.1+ MB


None

---------- 

Проверка на пустые значения 



sex                 0
SMK_stat_type_cd    0
dtype: int64

## Описание данных в датасете 

Для анализа используется датасет [Smoking and Drinking Dataset with Body Signal](https://www.kaggle.com/datasets/sooyoungher/smoking-drinking-dataset)

Датасет содержит медицинские показатели более чем 990 тысяч человек, включая информацию о поле, возрасте, антропометрических параметрах, результатах анализов крови, а также сведения о курении и употреблении алкоголя.
Для исследования используются два номинальных признака:

- sex — пол человека (Male, Female)
- SMK_stat_type_cd — курит ли человек

Причем в исходном наборе данных выделяется три категории курильщиков:
- 1.0 - не курит
- 2.0 - курит переодически
- 3.0 - постоянных курильщик

Поэтому в ходе исследования будет использоватна только группа 1.0 и 3.0, группа 2.0 будет удалена для частоты эксперимента.

In [31]:
group_male = df[(df['sex']== 'Male') & (df['SMK_stat_type_cd'].isin([1.0, 3.0]))][['sex', 'SMK_stat_type_cd']]
group_female = df[(df['sex']== 'Female') & (df['SMK_stat_type_cd'].isin([1.0, 3.0]))][['sex', 'SMK_stat_type_cd']]

# Замена объектов на категориальный тип данных, для корректности поставновки задачи
group_male['SMK_stat_type_cd'] = group_male['SMK_stat_type_cd'].map({1.0: 'N', 3.0: 'Y'})
group_female['SMK_stat_type_cd'] = group_female['SMK_stat_type_cd'].map({1.0: 'N', 3.0: 'Y'})


print('Проверка значений по group_male')
display(group_male.value_counts())
print('-'*10)
print('Проверка значений по group_female')
display(group_female.value_counts())

Проверка значений по group_male


sex   SMK_stat_type_cd
Male  Y                   197706
      N                   164681
Name: count, dtype: int64

----------
Проверка значений по group_female


sex     SMK_stat_type_cd
Female  N                   437760
        Y                    16248
Name: count, dtype: int64

In [15]:
group_male = df[df['sex']== 'Male'][['sex', 'DRK_YN']]
group_female = df[df['sex']== 'Female'][['sex', 'DRK_YN']]

print('Проверка значений по group_male')
display(group_male.value_counts())
print('-'*10)
print('Проверка значений по group_female')
display(group_female.value_counts())

Проверка значений по group_male


sex   DRK_YN
Male  Y         354317
      N         172098
Name: count, dtype: int64

----------
Проверка значений по group_female


sex     DRK_YN
Female  N         323760
        Y         141171
Name: count, dtype: int64

# Постановка статистической задачи и теоретические сведения

Цель исследования: проверка наличия статистиечскиой связи между полом человека и фактом курения: влияет ли пол на приверженность к курению. Оба принака измеряются в номинальной шкале. 

В рамках представленного набора данных я рассматриваю две величины

* sex — пол человека (Male/Female);
* SMK_stat_type_cd — статус курения (Y/N).

Для этого сформулировал следующие гипотезы:

$$
H_0: X \text{ и } Y \text{ независимы, т.е. вероятность принадлежности человека к группе курящих одинакова для мужчин и женщин} 
$$

$$
H_1: X \text{ и } Y \text{ гипотеза предполагает существование статистической зависимости между рассматриваемыми признаками}
$$

---

## Таблица сопряжённости

Для анализа двух номинальных признаков строится таблица сопряжённости, содержащая частоты наблюдений для всех возможных комбинаций значений признаков.

Общий вид таблицы:

|       | $$Y_1$$        | $$Y_2$$        | $$...$$ | $$Итого$$        |
| ----- | ------------ | ------------ | --- | ------------ |
| $$X_1$$ | $$O_{11}$$     | $$O_{12}$$     | $$...$$ | $$O_{1\cdot}$$ |
| $$X_2$$ | $$O_{21}$$     | $$O_{22}$$    | $$...$$ | $$O_{2\cdot}$$ |
| $$...$$   | $$...$$          | ...          | $$...$$ | $$...$$          |
| $$Итого$$ | $$O_{\cdot1}$$ | $$O_{\cdot2}$$ | $$...$$ | $$n$$          |

Где в таблице: 
$$
O_{ij} — наблюдаемая частота;
n — объём выборки.
$$

Если признаки независимы, то частоты в таблице должны быть близки к тем, которые ожидаются при независимом распределении признаков.

---

## Критерий хи-квадрат Пирсона

Для проверки гипотезы о независимости используется критерий хи-квадрат Пирсона, благодаря которому сначала вычисляются ожидаемые частоты:

$$
E_{ij}=
\frac{O_{i\cdot}\cdot O_{\cdot j}}{n}
$$

где:
$$
O_{i\cdot} — сумма наблюдений в (i)-й строке;
O_{\cdot j} — сумма наблюдений в (j)-м столбце;
n — общий объём выборки.
$$

Затем вычисляется статистика критерия:
$$
\chi^2=
\sum_{i=1}^{r}
\sum_{j=1}^{c}
\frac{(O_{ij}-E_{ij})^2}{E_{ij}}
$$

где:

* r — число строк таблицы сопряжённости;
* c — число столбцов таблицы сопряжённости.

Чем сильнее наблюдаемые частоты отличаются от ожидаемых, тем больше значение статистики **chi^2**, а следовательно, тем больше оснований считать признаки зависимыми.

При выполнении нулевой гипотезы статистика имеет распределение хи-квадрат с числом степеней свободы

$$
df=(r-1)(c-1)
$$

Для таблицы (2 \times 2):

$$
df=(2-1)(2-1)=1
$$

Для принятия решения задаю уровень значимости
$$
\alpha = 0.05.
$$

---

## Меры связи для номинальных признаков

Сам факт отклонения гипотезы ещё не показывает силу связи между признаками. Поэтому дополнительно рассчитываются специальные коэффициенты.

### Коэффициент Фи

Для таблиц размера 2X2 используется коэффициент Фи:

$$
\phi=
\sqrt{\frac{\chi^2}{n}}
$$

где:

* (\chi^2) — значение статистики Пирсона;
* (n) — объём выборки.

Коэффициент принимает значения от 0 до 1:

* 0 — связь отсутствует;
* 1 — максимально сильная связь.

---

### Коэффициент Крамера

Для таблиц произвольного размера применяется коэффициент V Крамера:

$$
V=
\sqrt{
\frac{\chi^2}
{n\cdot\min(r-1,c-1)}
}
$$

Коэффициент также изменяется в диапазоне от 0 до 1 и позволяет оценить тесноту связи между качественными признаками.

---

## Интерпретация силы связи

Для коэффициента Фи и коэффициента V Крамера часто используют следующую шкалу интерпретации:

| Значение коэффициента | Сила связи              |
| --------------------- | ----------------------- |
| 0 – 0.10              | практически отсутствует |
| 0.10 – 0.30           | слабая                  |
| 0.30 – 0.50           | умеренная               |
| более 0.50            | сильная                 |

Таким образом, в данной работе сначала будет выполнена проверка гипотезы о независимости пола и факта курения с помощью критерия хи-квадрат Пирсона, после чего сила обнаруженной связи будет оценена с использованием коэффициента Фи и коэффициента V Крамера. Это позволит не только определить наличие статистической зависимости, но и оценить её практическую значимость.


In [36]:
# Построение таблицы сопряженности 
df_for_table = pd.concat([group_female, group_male])
table = pd.crosstab(df_for_table['sex'], df_for_table['SMK_stat_type_cd'], margins=True)

print("Таблица сопряженности: \n")
print(table)

Таблица сопряженности: 

SMK_stat_type_cd       N       Y     All
sex                                     
Female            437760   16248  454008
Male              164681  197706  362387
All               602441  213954  816395


In [43]:
# Наблюдаемые частоты  
O = np.array([
    [164681, 197706],
    [437760, 16248]
])

n = O.sum()
row_totals = O.sum(axis=1)
col_totals = O.sum(axis=0)
print("Объем выборки:", n)

# Ожидаемые частоты
E = np.zeros(O.shape)

for i in range(O.shape[0]):
    for j in range(O.shape[1]):
        E[i, j] = row_totals[i] * col_totals[j] / n

print("\nОжидаемые частоты:")
print(E)

# Расчет χ² 
chi_square = 0

for i in range(O.shape[0]):
    for j in range(O.shape[1]):
        term = (O[i, j] - E[i, j])**2 / E[i, j]
        chi_square += term

        print(
            f"Ячейка ({i+1},{j+1}): "
            f"({O[i,j]} - {E[i,j]:.2f})² / {E[i,j]:.2f} "
            f"= {term:.4f}"
        )
print(f"\nχ² = {chi_square:.4f}")

r, c = O.shape
df = (r - 1) * (c - 1)
print(f"Степени свободы = {df}")

alpha = 0.05
chi_crit = chi2.ppf(1 - alpha, df)
print(f"Критическое значение χ² = {chi_crit:.4f}")

p_value = 1 - chi2.cdf(chi_square, df)
print(f"p-value = {p_value:.10f}")

phi = np.sqrt(chi_square / n)
print(f"\nКоэффициент Фи = {phi:.4f}")

cramers_v = np.sqrt(
    chi_square / (n * min(r - 1, c - 1))
)
print(f"Крамера = {cramers_v:.4f} \n")

print('='*10)

if chi_square > chi_crit:
    print("\nОтвергаем H0.")
    print("Пол и факт курения статистически зависимы.")
else:
    print("\nНет оснований отвергать H0.")
    print("Статистически значимая связь не обнаружена.")

Объем выборки: 816395

Ожидаемые частоты:
[[267415.63418076  94971.36581924]
 [335025.36581924 118982.63418076]]
Ячейка (1,1): (164681 - 267415.63)² / 267415.63 = 39468.1676
Ячейка (1,2): (197706 - 94971.37)² / 94971.37 = 111132.4974
Ячейка (2,1): (437760 - 335025.37)² / 335025.37 = 31503.3014
Ячейка (2,2): (16248 - 118982.63)² / 118982.63 = 88705.4244

χ² = 270809.3907
Степени свободы = 1
Критическое значение χ² = 3.8415
p-value = 0.0000000000

Коэффициент Фи = 0.5759
Крамера = 0.5759 


Отвергаем H0.
Пол и факт курения статистически зависимы.


# Результаты и обсуждение

Полученный результат можно объяснить как социальными, так и культурными особенностями исследуемой выборки.
Из таблицы сопряжённости видно, что среди мужчин доля курящих значительно выше, чем среди женщин. Такая ситуация характерна для многих стран мира, особенно для стран Азии, где исторически курение среди мужчин является более распространённым и социально приемлемым поведением.
Одной из причин может быть различие в социальных нормах и моделях поведения мужчин и женщин. Мужчины чаще вовлекаются в практики, связанные с употреблением табака, а также подвергаются большему влиянию соответствующих социальных факторов. Для женщин, напротив, курение во многих обществах встречается реже вследствие культурных ограничений и более негативного отношения со стороны общества.

Следует отметить, что очень большое значение статистики χ² связано не только с реальным различием между группами, но и с чрезвычайно большим объёмом выборки: **n=816395**

При таком количестве наблюдений даже относительно небольшие различия между группами становятся статистически значимыми. Поэтому наряду с проверкой гипотезы важно анализировать меры силы связи. В данном исследовании коэффициенты Фи и Крамера также показали достаточно сильную связь, что подтверждает практическую значимость полученного результата.
Таким образом, проведённый анализ позволяет заключить, что в рассматриваемой выборке вероятность того, что человек является курильщиком, существенно связана с его полом, причём среди мужчин курение встречается значительно чаще, чем среди женщин.

